File: packages/python-sdk/src/services/proposal_service.py

from google.cloud import pubsub_v1
from typing import Any, Callable, Optional
from web3 import Contract
from eth_typing import Address
from ..types import Proposal

class ProposalService:
    def __init__(self, project_id: str, topic_name: str, task_registry: Contract):
        self.pubsub = pubsub_v1.PublisherClient()
        self.topic_name = topic_name
        self.task_registry = task_registry
        self.subscription = None
        self.on_new_proposal: Optional[Callable[[Proposal], None]] = None
        self.project_id = project_id
        
    async def setup_subscription(self, user_id: str):
        subscription_name = f"tasks-{user_id}"
        topic_path = self.pubsub.topic_path(self.project_id, self.topic_name)
        
        # Create topic if it doesn't exist
        try:
            self.pubsub.get_topic(request={"topic": topic_path})
        except Exception:
            self.pubsub.create_topic(request={"name": topic_path})
            
        subscription_path = self.pubsub.subscription_path(
            self.project_id, subscription_name
        )
        
        # Create subscription if it doesn't exist
        try:
            self.pubsub.get_subscription(request={"subscription": subscription_path})
        except Exception:
            self.pubsub.create_subscription(
                request={"name": subscription_path, "topic": topic_path}
            )
            
        def callback(message):
            if self.on_new_proposal:
                proposal = Proposal.from_dict(message.data)
                self.on_new_proposal(proposal)
            message.ack()
            
        subscriber = pubsub_v1.SubscriberClient()
        self.subscription = subscriber.subscribe(subscription_path, callback)
        
    async def send_proposal(self, task_id: str, agent_id: str, price: int) -> None:
        topic_path = self.pubsub.topic_path(self.project_id, self.topic_name)
        data = {
            "taskId": task_id,
            "price": str(price),
            "agent": agent_id
        }
        
        try:
            future = self.pubsub.publish(
                topic_path, 
                str(data).encode("utf-8")
            )
            future.result()
            print(f"Proposal for task {task_id} sent successfully.")
        except Exception as e:
            print(f"Failed to send proposal for task {task_id}:", e)
            raise e
            
    async def get_proposals(self, task_id: str) -> list[str]:
        # Implement PubSub message pulling if needed
        return []
        
    def set_on_new_proposal_listener(self, listener: Callable[[Proposal], None]):
        self.on_new_proposal = listener
        
    async def approve_proposal(self, task_id: int, proposal: Proposal) -> None:
        try:
            tx = await self.task_registry.functions.approveProposal(
                task_id, 
                proposal
            ).transact()
            await tx.wait()
        except Exception as e:
            print("Approving proposal failed:", e)
            raise e

File: packages/sdk/src/services/ProposalService.ts
import { PubSub } from '@google-cloud/pubsub';
import { Proposal } from '../types';
import { BigNumberish, ethers } from 'ethers';

export interface ProposalServiceParams {
  projectId: string;
  topicName: string;
  taskRegistry: ethers.Contract;
}

export class ProposalService {
  private pubsub: PubSub;
  private topicName: string;
  private subscription: any;
  private taskRegistry: ethers.Contract;
  protected onNewProposal: (proposal: Proposal) => void = () => {};

  constructor(params: ProposalServiceParams) {
    this.pubsub = new PubSub({projectId: params.projectId});
    this.topicName = params.topicName;
    this.taskRegistry = params.taskRegistry;
  }


  /**
   * Subscribes to new proposals using Google PubSub.
   */
  public async setupSubscription(userId: string) {
    const subscriptionNameOrId = `tasks-${userId}`;

    // Create a topic if it doesn't exist
    const topic = this.pubsub.topic(this.topicName);
    
    // Check if the subscription already exists
    const [subscriptions] = await topic.getSubscriptions();
    this.subscription = subscriptions.find(sub => sub.name.endsWith(subscriptionNameOrId));
    
    if (!this.subscription) {
      // Create a subscription to the topic if it doesn't exist
      [this.subscription] = await topic.createSubscription(subscriptionNameOrId);
      console.log(`Subscription ${this.subscription.name} created.`);
    } else {
      console.log(`Subscription ${this.subscription.name} already exists.`);
    }

    // Create an event handler to handle messages
    let messageCount = 0;
    const messageHandler = (message: any) => {
      console.log(`Received message ${message.id}:`);
      const proposal = JSON.parse(message.data);
      console.log(`\Proposal: ${proposal}`);
      console.log(`\tAttributes: ${message.attributes}`);
      messageCount += 1;
      this.onNewProposal(proposal);
      // "Ack" (acknowledge receipt of) the message
      message.ack();
    };
    // Listen for new messages until timeout is hit
    this.subscription.on('message', messageHandler);
  }

  /**
   * Sends a proposal for a task.
   * @param {string} taskId - The ID of the task.
   * @param {BigNumberish} price - The price of the proposal.
   * @returns {Promise<void>} A promise that resolves when the proposal is sent.
   */
  async sendProposal(taskId: string, agentId: string, price: BigNumberish): Promise<void> {
    const pubsub = new PubSub();
    const topicName = this.topicName;
    console.log(`Task ID: ${taskId} (type: ${typeof taskId})`);
    console.log(`Agent ID: ${agentId} (type: ${typeof agentId})`);
    console.log(`Price: ${price} (type: ${typeof price})`);
    const data = JSON.stringify({
      taskId: taskId.toString(),
      price: price.toString(),
      agent: agentId
    });

    const dataBuffer = Buffer.from(data);

    try {
      await pubsub.topic(topicName).publishMessage({ data: dataBuffer });
      console.log(`Proposal for task ${taskId} sent successfully.`);
    } catch (error) {
      console.error(`Failed to send proposal for task ${taskId}:`, error);
      throw error;
    }
  }

  /**
   * Gets proposals for a task.
   * @param {string} taskId - The ID of the task.
   * @returns {Promise<string[]>} A promise that resolves to an array of proposal IDs.
   */
  async getProposals(taskId: string): Promise<string[]> {
    // USE PUBSUB
    return [];
  }
  

  /**
   * Sets the listener for new proposals.
   * @param listener - The function to be called when a new proposal is created.
   */
  setOnNewProposalListener(listener: (proposal: Proposal) => void) {
    this.onNewProposal = listener;
  }

  /**
   * Approves a proposal for a task.
   * @param {BigNumberish} taskId - The ID of the task.
   * @param {Object} proposal - The proposal object.
   * @param {BigNumberish} proposal.id - The ID of the proposal.
   * @param {BigNumberish} proposal.price - The price of the proposal.
   * @param {BigNumberish} proposal.taskId - The ID of the task.
   * @param {string} proposal.agent - The agent address.
   * @returns {Promise<void>} A promise that resolves when the proposal is approved.
   */
  async approveProposal(taskId: BigNumberish, proposal: { id: BigNumberish; price: BigNumberish; taskId: BigNumberish; agent: string }): Promise<void> {
    try {
      const tx = await this.taskRegistry.approveProposal(taskId, proposal);
      await tx.wait();
    } catch (error) {
      console.error("Approving proposal failed:", error);
      throw error;
    }
  }
} 